pkg/api/resource/quantity.go

```diff
@@ -372,41 +378,53 @@ func NewMilliQuantity(value int64, format Format) *Quantity {
 	}
 }
 
+// NewScaledQuantity returns a new Quantity representing the given
+// value * 10^scale in DecimalSI format.
+func NewScaledQuantity(value int64, scale Scale) *Quantity {
+	return &Quantity{
+		Amount: inf.NewDec(value, inf.Scale(-scale)), // inf.Scale is upside-down
+		Format: DecimalSI,
+	}
+}
+
 // Value returns the value of q; any fractional part will be lost.
 func (q *Quantity) Value() int64 {
-	if q.Amount == nil {
-		return 0
-	}
-	tmp := &inf.Dec{}
-	return tmp.Round(q.Amount, 0, inf.RoundUp).UnscaledBig().Int64()
+	return q.ScaledValue(0)
 }
 
-// MilliValue returns the value of q * 1000; this could overflow an int64;
+// MilliValue returns the value of ceil(q * 1000); this could overflow an int64;
 // if that's a concern, call Value() first to verify the number is small enough.
 func (q *Quantity) MilliValue() int64 {
+	return q.ScaledValue(Milli)
+}
+
+// ScaledValue returns the value of ceil(q * 10^scale); this could overflow an int64;
+func (q *Quantity) ScaledValue(scale Scale) int64 {
 	if q.Amount == nil {
 		return 0
 	}
+	infScale := inf.Scale(-scale) // inf.Scale is upside-down
 	tmp := &inf.Dec{}
-	return tmp.Round(tmp.Mul(q.Amount, decThousand), 0, inf.RoundUp).UnscaledBig().Int64()
+	return tmp.Round(q.Amount, infScale, inf.RoundUp).UnscaledBig().Int64()
 }
 
 // Set sets q's value to be value.
 func (q *Quantity) Set(value int64) {
-	if q.Amount == nil {
-		q.Amount = &inf.Dec{}
-	}
-	q.Amount.SetUnscaled(value)
-	q.Amount.SetScale(0)
+	q.SetScaled(value, 0)
 }
 
 // SetMilli sets q's value to be value * 1/1000.
 func (q *Quantity) SetMilli(value int64) {
+	q.SetScaled(value, Milli)
+}
+
+// SetScaled sets q's value to be value * 10^scale
+func (q *Quantity) SetScaled(value int64, scale Scale) {
 	if q.Amount == nil {
 		q.Amount = &inf.Dec{}
 	}
 	q.Amount.SetUnscaled(value)
-	q.Amount.SetScale(3)
+	q.Amount.SetScale(inf.Scale(-scale)) // inf.Scale is upside-down
```

Comment:
optional suggestion: You could make a `func (s Scale) infScale() inf.Scale` to ensure this logic is done consistently everywhere...